arrows/vxl/high_pass_filter.cxx

```diff
@@ -0,0 +1,489 @@
+    box_average_horizontal( grey_img, filter_xy, kernel_width );
+    vertical_box_bidirectional_pass( grey_img, filter_xy, filter_y,
+                                     kernel_height );
+    vil_math_image_max( filter_x, filter_y, filter_xy );
+
+    return output;
+  }
+
+  // Function which returns an image view of the even rows of an image
+  template < typename PixType >
+  inline vil_image_view< PixType >
+  even_rows( const vil_image_view< PixType >& im )
+  {
+    return vil_image_view< PixType >( im.memory_chunk(), im.top_left_ptr(),
+                                      im.ni(), ( im.nj() + 1 ) / 2,
+                                      im.nplanes(),
+                                      im.istep(), im.jstep() * 2,
+                                      im.planestep() );
+  }
+
+  // Function which returns an image view of the odd rows of an image
+  template < typename PixType >
+  inline vil_image_view< PixType >
+  odd_rows( const vil_image_view< PixType >& im )
+  {
+    return vil_image_view< PixType >( im.memory_chunk(),
+                                      im.top_left_ptr() + im.jstep(),
+                                      im.ni(), ( im.nj() ) / 2, im.nplanes(),
+                                      im.istep(), im.jstep() * 2,
+                                      im.planestep() );
+  }
+
+  // Fast 1D (horizontal) box filter smoothing
+  template < typename PixType >
+  void
+  box_average_horizontal( const vil_image_view< PixType >& src,
+                          vil_image_view< PixType >& dst,
+                          unsigned kernel_width )
+  {
+    if( src.ni() <= 0 )
+    {
+      LOG_ERROR( p->logger(), "Image width must be non-zero" );
+    }
+    if( kernel_width % 2 == 0 )
+    {
+      LOG_ERROR( p->logger(), "Kernel width must be odd" );
+    }
+
+    if( kernel_width >= src.ni() )
+    {
+      // Force width to be smaller than image width and odd
+      kernel_width = ( src.ni() == 1 ? 1 : ( src.ni() - 2 ) | 0x01 );
+    }
+
+    unsigned ni = src.ni(), nj = src.nj(), np = src.nplanes();
+    dst.set_size( ni, nj, np );
+
+    unsigned half_width = kernel_width / 2;
+
+    std::ptrdiff_t istepS = src.istep(),  jstepS = src.jstep(),
+                   pstepS = src.planestep();
+    std::ptrdiff_t istepD = dst.istep(),  jstepD = dst.jstep(),
+                   pstepD = dst.planestep();
+
+    const PixType*   planeS = src.top_left_ptr();
+    PixType*         planeD = dst.top_left_ptr();
+
+    for( unsigned p = 0; p < np; ++p, planeS += pstepS, planeD += pstepD )
+    {
+      const PixType* rowS = planeS;
+      PixType*       rowD = planeD;
+
+      for( unsigned j = 0; j < nj; ++j, rowS += jstepS, rowD += jstepD )
+      {
+        const PixType* pixelS1 = rowS;
+        const PixType* pixelS2 = rowS;
+        PixType*       pixelD = rowD;
+
+        // fast box filter smoothing by adding one pixel to the sum and
+        // subtracting another pixel at each step
+        unsigned sum = 0;
```

Comment:
Okay... I'm not immediately sure what to do with this (might want to ask the original author, is that Dawkins?), but this is triggering float-to-int warnings that are concerning, as in "this might be properly broken". However, I also don't think we want to just use `PixType` because I get the impression that could lead to overflow and would be equally broken. I wonder if perhaps this needs some finessing to choose an appropriate intermediate type.

Reply:
Yes, I believe @mattdawkins wrote the majority of this. I assume using a `double` is just inefficient and could have accuracy issues for integral types? I'm going to look at VXL to see if there's either a drop-in replacement or something that can be used as a reference. 

Also, this might be a bit of a cop-out but at least for burn-out this should only take `VXL_BYTE` images as input.

Reply:
I was thinking of using `double` for any floating point and `int64` otherwise. We might want to remove support for 64-bit (integer) input, but the others should be passably okay with such an approach.

Reply:
Burn-Out (and specifically this adaboost approach) is heavily optimized for and has really only ever had 8-bit RGB inputs (though greyscale was supported as well in old as well). You might just want to add an error early on checking for this as opposed to over-optimizing for cases we don't care about, for filters which require it. There might be one other.